javascript/react-js/react-testing-part-two.md

```diff
@@ -0,0 +1,155 @@
+
+### Learning Outcomes
+
+By the end of this lesson, you should be able to do the following:
+
+TODO
+
+### What is Mocking?
+
+If you've been following along our lessons, the concept of mocking has already been introduced in the earlier section and you might have even incorporated mocks in your [Battleship](https://www.theodinproject.com/paths/full-stack-javascript/courses/javascript/lessons/battleship) project. Let's look at how mocks'll help in testing React components.
+
+#### Testing Callback Handlers
+
+Callbacks are ubiquitous. Every avenue of user interaction involves callbacks. Sometimes they're passed in as props to alter state of the parent component. Consider this simple input component:
+
+~~~javascript
+// FavoriteInput.js
+
+import React from "react";
+
+const FavoriteInput = ({ onChange, id }) => (
+  <label htmlFor={id}>
+    What is your favorite wild animal?
+    <input id={id} onChange={onChange} />
+  </label>
+);
+
+export default FavoriteInput;
+~~~
+
+Nothing fancy. `FavoriteInput` is a simple component with a couple props passed in. We're interested in the `onChange` prop. We have no idea what the function does. We have no idea how the function will affect the application. All we know is it must be called when user types in the input box. Let's test it. 
+
+~~~javascript
+// FavoriteInput.test.js
+
+import React from "react";
+import { render } from "@testing-library/react";
+import "@testing-library/jest-dom";
+import userEvent from "@testing-library/user-event";
+import FavoriteInput from "./FavoriteInput";
+
+describe("Favorite Input", () => {
+  let onChange;
+  let getByRole;
+  let input;
+
+  beforeEach(() => {
+    onChange = jest.fn();
+    ({ getByRole } = render(<FavoriteInput onChange={onChange} />));
+    input = getByRole("textbox");
+  });
+
+  test("calls onChange correct number of times", () => {
+    userEvent.type(input, "Lion");
+    expect(onChange).toHaveBeenCalledTimes(4);
+  });
+
+  test("calls onChange with correct argument", () => {
+    userEvent.type(input, "Ox");
+
+    expect(onChange.mock.calls[0][0].target.value).toMatch("O");
+    expect(onChange.mock.calls[1][0].target.value).toMatch("x");
+  });
+
+  test("input has correct values", () => {
+    userEvent.type(input, "Whale");
+    expect(input).toHaveValue("Whale");
+  });
+});
+~~~
+
+Three tests and we are done with this component. Take some time to figure out what functions come from which package. 
+We mock the `onChange` handler using one of jest features, `jest.fn()`. For the first test, we assert that the mock function is invoked correct number of times. While the second test ensures that the mock function is called with the correct arguments. The third test seems redundant, and it is; it's just here to show what other ways we could've tested the component. We use `beforeEach` to clump all the variable declarations in one place rather than defining them individually for each test.  
```

Comment:
I also think that we should talk about mocking implementation in this lesson somewhere.

Reply:
> I also think that we should talk about mocking implementation in this lesson somewhere.

Sure. Where do you think in the lesson would a discussion about this be most suitable? Also, I'm wondering how much should we cover this? Would a couple of brief sentences suffice?

Reply:
I'm not entirely sure where would be best. Considering the scope of this lesson I actually wonder if it should go in your part one lesson. Do you have any thoughts on that idea?

As far as how much of a lesson to spend talking about this I think one short paragraph or less would be able to cover things in good enough depth.

Reply:
> Do you have any thoughts on that idea?

I do prefer having it here since this is the lesson where the word "mocking" comes up. I'm fine either way tho.

-----------------------------------------------


https://kentcdodds.com/blog/testing-implementation-details

^ I found this clean article that talks about testing implementation thoroughly (how it's detrimental to test implementation). I suggest we add it as an assignment. With this article in this lesson, we might not even need a section on "mocking implementation".